### ISSUE TITLE
Orleans clients using PersistentStreamProviders must have configured gateways for every silo in the cluster.

### ISSUE THREAD

**Issue by jason-bragg** [dotnet/orleans#1069]
Orleans clients communicate with the cluster through configured gateways.
http://dotnet.github.io/orleans/Orleans-Configuration-Guide/Client-Configuration.
If a grain activation is on a silo that is not on one of the configured gateways, Orleans will route communication with that grain through one of the configured gateways.  This is possible because each gateway can obtain the address of the silo responsible for any activation.

System targets, however, are not routed, as their address cannot be obtained.  Any call to a system target must be made directly to the silo responsible for that system target's activation.  This means that an Orleans client that needs to send a message to a system target must have the silo responsible for the system target's activation in it's list of configured gateways.

Since the Persistent Stream Providers used system targets to read from the queue, for Orleans Clients to use Persistent Stream Providers they must have gateway connections to every silo in the cluster.

Orleans clients should require only a single gateway.  Since this breaks that expectation, this is a bug.

**Comment by gabikliot** [comment]
Hmmm, interesting.
It sounds like there is indeed a bug around that area. But may I propose a different fix from what you suggested.
You suggested as a fix that a client has to be connected to every GW to be able to send request to any system target. I think we can achieve that differently/better.

When client sends a msg to a system target, it has the system target address, including silo address (system target address in the streaming case comes from the producer Grain Reference returned by the pub sub). So what you wrote: "System targets, however, are not routed, as their address cannot be obtained. " **is not correct**. GW doesn't need to use the directory to route to system target, obviously, but it has the system target address in the msg header. So it can route the msg to the correct silo.

The bug is that most probably the system target silo address is not copied somewhere/gets lost. We just need to make sure it is not lost.

Regardless of how we fix it, requiring the client to be connected to all GWs is wrong - the system was designed, specifically, to allow the client to be connected to a subset, even one, GW and everything should work correctly.

If you agree with all said above, I can help search through the code t find out where we are loosing that system target address.

**Comment by gabikliot** [comment]
OK, so the problem is on that line:
https://github.com/dotnet/orleans/blob/master/src/Orleans/Messaging/ProxiedMessageCenter.cs#L178

A msg to a system target where TargetSilo is specified can go only to this GW. Otherwise, we just drop it. The bug is that we just drop it.
The fix I recommend is:
first check sg.TargetSilo != null and try to send directly to the address GW. But in the else case, if gatewayConnections.TryGetValue returns false or is not alive, just raise a boolean flag `notYetSent`.
Then, if `notYetSent` do the other 2 cases: msg.TargetGrain.IsSystemTarget || msg.IsUnordered or else.
So basically, in this case we will send this msg to a round robin GW.

I think that will be enough. There is a chance we will later overwrite the original TargetSilo with GW silo address, which will break this. But I think we are fine, due to that line: https://github.com/dotnet/orleans/blob/master/src/Orleans/Messaging/GatewayConnection.cs#L214
But if we are not fine, and we do overwrite the original TargetSilo with GW silo address, we should just find where and change it.

**Comment by jason-bragg** [comment]
>You suggested as a fix that a client has to be connected to every GW to be able to send request to any system target. I think we can achieve that differently/better.

That wasn't a proposed solution.  The fact that this is necessary is part of the problem.  I added that bit to help elaborate where this issue was coming from, since system targets are internal to Orleans and clients generally have no need of them, which is probably why this issue hasn't come up before.

 >"System targets, however, are not routed, as their address cannot be obtained. " is not correct

I think you're being a little nit picky here :)  The code that routes grain calls, by getting their location from the directory, can't get the address for system targets.  I was attempting to describe the problem without going into too much details regarding how the routing was done.  Yes, we _can_ route them, because they're addressed, but that's different routing logic, which does not exist.

I've not spent much time on this, but I think you're on the right path.  Having said that, we have special case logic for system targets in several places, like those you noted and:
https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/Messaging/GatewayAcceptor.cs#L103
I think we'll need to review all of these cases and make sure they support routing.

**Comment by gabikliot** [comment]
OK, lets not argue about who suggested and implied what. :-) I am merely trying to help, and provide more in depth description of how different components work, mostly to pass the knowledge.

I believe routing to system target (not via directory of course, but a direct routing) is already enabled and works. When silo A gets a msg to a system target on Silo B, it know to forward that msg to Silo B. There might be bugs, of course, but in general that logic exists and works. I am pretty sure.

I do believe the fix I suggested will be enough. Please try it.

The line you pointed in the GWAcceptor is NOT a problem, due to a previous check: ` if (targetAddress == null)` https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/Messaging/GatewayAcceptor.cs#L97 . So the line 103 applies only when system target is address is not specified. Indeed, in that case we route to myself. But that is not our case.

**Comment by jason-bragg** [comment]
>I do believe the fix I suggested will be enough. Please try it.

Yup, going to look into it on the morrow.  Thanks!

> if (targetAddress == null)  

I may have misunderstood what I was looking at, but I thought that target was set only if we were routing a reply back to a client we were proxying for.  Which meant it would always be null for a system target, causing the gateway host to always try to handle the call to the system target itself.  I'm still a bit fuzzy on how the forwarding and proxying work. :/  I'll figure it out though..

**Comment by gabikliot** [comment]
You are right. That may indeed be that other place that I wrote: 
"There is a chance we will later overwrite the original TargetSilo with GW silo address, which will break this. ... But if we are not fine, and we do overwrite the original TargetSilo with GW silo address, we should just find where and change it."

So this does look like this other place.
The re-routing the works correctly from silo to silo, but not from client to silo in case of a system target.

The fix can be just to check `if (targetAddress == null)` and before we remove `Message.Header.TARGET_SILO` that `if(msg.TargetGrain.IsSystemTarget) and `if (msg.TargetSilo != null)` then keep the specified  msg.TargetSilo and just call `MessageCenter.RerouteMessage(msg);`

**Comment by jason-bragg** [comment]
@NingnaD is porting over the ClientAddressableTests from VSO.  I'll add a test for this to that after it's ported, then address this.
If a fix is needed sooner, I'll raise the priority.

**Comment by gabikliot** [comment]
:+1:

**Comment by amamh** [comment]
Maybe this is related to https://github.com/dotnet/orleans/issues/1031

**Comment by jason-bragg** [comment]
@amamh Looking at your configurations in #1031, this would explain the intermittent failure you are seeing.  If you explicitly specify all the gateways in your client configurations, does the intermittent failure go away?

**Comment by amamh** [comment]
@jason-bragg 
Yes, that fixed the issue. Am I to understand that is a bug and I won't need to do that in the future when it's fixed?

**Comment by jason-bragg** [comment]
>Am I to understand that is a bug and I won't need to do that in the future when it's fixed?

@amamh, Correct.  Orleans clients should require only a single gateway.  This bug was exposed by the fact that the persistent stream provider uses a system target to communicate with clients, which is not a common pattern.  To my knowledge, this is only an issue while using the persistent stream provider.

Also, to address the ambiguity in the description of this problem, resulting in confusion regarding suggested solutions (see @gabikliot's comments), I'll update the description to clarify.

**Comment by gabikliot** [comment]
@jason-bragg , now that ClientAddressableTests  from VSO have been ported to GH, maybe we should fix this issue? I think the fix will be trivial, a couple of lines that we already discussed above, and you have the test already.

**Comment by benjaminpetit** [comment]
Fixed with #4254
